pkg/kubelet/controller.go

```diff
@@ -155,6 +155,18 @@ func (c *Controller) nodeAddress(node v1.Node) (string, map[v1.NodeAddressType][
 	return "", m, fmt.Errorf("host address unknown")
 }
 
+// nodeIsReady checks the node for the Ready condition to see if it is set to
+// "True". Without this condition status, we cannot guarantee the node is
```

Comment:
That's a great question, and I think my logic needs to slightly change. I dug through Kubernetes source code and here are my findings.

From a controller standpoint, `kube-controller-manager` (specifically the `node-lifecycle-controller`) is mostly passive, just collecting what nodes' kubelet instances report to it for status and condition (and _if_ they report to it, documented below). It's essentially collecting and making a final source of truth determination for node status.

The bulk of the work is happening on the nodes themselves, within kubelet. The following are what kubelet does for checks:

### Kubelet checks

* Runtime errors
	* Container runtime status check hasn't completed
	* Container runtime is down
	* Pod lifecycle event generator is down
* Network errors
	* Container runtime network not ready
* Storage errors
	* CSINode (CSI provider on the node) is not initialized
* Node shutdown manager
	* Node is shutting down
* Missing required node capacities
	* CPU
	* Memory
	* Pods (max pods per node)

Based off of the routine collection of the above, kubelet will report the following **node ready status** to the control plane.

### Node Ready status

* **True** - All kubelet checks succeed and return green (no errors, no shutdown, all required capacities present)
* **False** - Any of the kubelet checks are not green (existing errors, and/or shutdown, and/or missing required capacities)
* **Unknown** - `kube-controller-manager` hasn't heard from the node's kubelet in the grace period (40 seconds)

Because of this new information, I think we should continue to scrape a target even if `NodeReady` status is set to `ConditionFalse` (kubelet is communicating with the control plane, but something isn't right). This would be a change do this PR's logic. But we should *not* be scraping if `NodeReady` status is `ConditionUnknown` (kubelet hasn't sent anything in grace_period+ seconds).

So I think `nodeIsReady` should be changed to:

```go
func nodeIsReady(node v1.Node) bool {
	for _, c := range node.Status.Conditions {
		if c.Type == v1.NodeReady && c.Status != v1.ConditionUnknown {
			return true
		}
	}
	return false
}
```

Instead of `c.Status == v1.ConditionTrue`. This function should also be renamed to something that reflects the check more accurately, probably `nodeReadyConditionKnown` or even something a little better than that.

Reply:
To answer your original comment:

> It can be very valuable to collect data from unhealthy but reachable nodes.

Yes, I agree.

Reply:
I have made these changes, but I'll keep this thread open and unresolved if you have any additional feedback on this.